Orange/misc/server_embedder.py

```diff
@@ -153,32 +158,80 @@ async def embedd_batch(
         EmbeddingCancelledException:
             If cancelled attribute is set to True (default=False).
         """
-        requests = []
+        # in Orange 3.33 keep content of the if - remove if clause and complete else
+        if processed_calback is None:
+            progress_items = iter(linspace(0, 1, len(data)))
+
+            def success_callback():
+                """Callback called on every successful embedding"""
+                callback(next(progress_items))
+
+        else:
+            warnings.warn(
+                "process_callback is deprecated and will be removed in version 3.33, "
+                "use callback instead",
+                FutureWarning,
+            )
+            success_callback = processed_calback
+
+        results = [None] * len(data)
+        queue = asyncio.Queue()
+
+        # fill the queue with items to embedd
+        for i, item in enumerate(data):
+            queue.put_nowait(TaskItem(id=i, item=item, no_repeats=0))
+
         async with AsyncClient(
-                timeout=self.timeout, base_url=self.server_url, proxies=get_proxies()
+            timeout=self.timeout, base_url=self.server_url, proxies=get_proxies()
         ) as client:
-            for p in data:
-                if self._cancelled:
-                    raise EmbeddingCancelledException()
-                requests.append(self._send_to_server(p, client, proc_callback))
+            tasks = self._init_workers(client, queue, results, success_callback)
 
-            embeddings = await asyncio.gather(*requests)
-        self._cache.persist_cache()
-        assert self.num_parallel_requests == 0
+            # wait for the queue to complete or one of workers to exit
+            queue_complete = asyncio.create_task(queue.join())
+            await asyncio.wait(
```

Comment:
I am not sure I completely understand this stopping criterion. Just the double check the intent on this wait: you wait for completion OR for one of the tasks to raise an EmbeddingConnectionError and thus crash?